### ISSUE TITLE
Field report v8: first run on a new target (transformers#48748) — grep found a precedent that exists in no thread, and `symbol` answered a question I did not ask

### ISSUE THREAD

**Issue by tarekziade** [huggingface/relore#58]
Follow-up to #8, same format, but the first one that is **not** a re-run. Every report in that thread (v0, v2, v3, v7) targeted `huggingface/transformers#48630`, so each measured what the last round of fixes changed for a reader who already had a known-good path through the tool. This one points the same agent at a bug it has never seen, to test the thing the re-runs could not: does the tool generalise, or had six reports quietly tuned it to one issue?

**Task:** diagnose `huggingface/transformers#48748` — *"`DogeForCausalLM` attends to future tokens with the default `sdpa` backend"*.

**Agent:** Claude Opus 5 via Claude Code. Cold on the bug; carrying knowledge of relore from the v7 session, so this is not a cold read of the tool.

**Result:** root cause confirmed from source, the in-flight PR found on call three, and **two findings neither the issue nor the PR contains** — a second broken backend, and a structural reason the proposed fix cannot be written where the PR writes it. No clone. ~49 server calls.

---

## Headline: the code lens is a discovery surface, not just a confirmation surface

v7's thesis was `--help`'s own line — *"a `[contributor claim]` confirmed by grep is stronger evidence than either alone"* — and it held again here. But the highest-value moment of this session inverts it, and I think it is the more important claim.

The bug is: Doge's attention always builds its own dense mask and passes it to the attention interface, so sdpa never gets `is_causal=True`, while `create_causal_mask` has already returned `None` under the sdpa skip. Nothing applies causality. The question that decides the review is: **has anyone in transformers already solved this shape, and how?**

`search` cannot answer it. Three honestly-phrased queries:

```console
$ relore search "model builds its own attention mask must force allow_is_causal_skip False"
10 hits  →  #41856, #41834, #46836, #36625, #47243 … all adjacent, none the precedent

$ relore search "indexer sparse attention causality force mask creation"
0 hits

$ relore search "doge dynamic mask attention sdpa causal"
2 hits  →  #48748 and #48769.  Correct, and exactly the two threads I already had.
```

`grep` answers it in one call:

```console
$ relore grep 'allow_is_causal_skip=False'
src/transformers/masking_utils.py:1510  "deepseek_sparse_attention": partial(create_causal_mask, allow_is_causal_skip=False),
src/transformers/masking_utils.py:1512  "qwen_sparse_attention":     partial(create_causal_mask, allow_is_causal_skip=False),
src/transformers/models/falcon/modeling_falcon.py:769  allow_is_causal_skip=False,
-- 8 hits in 4 of 6474 files searched
```

and the canonical instance, two calls later, is `AXK2Model.forward` — structurally identical to Doge, and correct:

```python
"allow_is_causal_skip": False,  # Always force creation to account for causality in the indexer
```

**That comment is the whole answer, and it exists only in code.** No thread argues it; it was written once and never discussed. The prose index is a faithful record of what people *said*, and the house rule for this bug class was never said — it was encoded. `search`'s `0 hits` on S2 is not a defect, it is the corpus being honest.

So the generalisation I would add to the v7 thesis:

> **Prose answers "what was decided". Code answers "what is done here".** For *"has this shape been solved before?"* — the question that decides most reviews — the code lens is the primary surface and the prose index is the fallback, not the other way round. `--help`'s TYPICAL ORDER puts the code verbs at step 5, as the check on step 3. On this session they were step 2.

I am not suggesting reordering `TYPICAL ORDER`, which is right for triage. I am suggesting the code verbs deserve a second entry point in it: *"step 5b — before proposing a fix, grep for how the fix is spelled elsewhere."*

---

## Two new uses for verbs that already shipped

**`copies` as fork detection.** I used it in v7 to group definitions. Here it did something else in one call:

```console
$ relore copies flex_attention_forward
3 definitions of flex_attention_forward in 2 shapes
-- shape 1: 2 copies (the majority shape)
   src/transformers/models/doge/modeling_doge.py:202
   src/transformers/models/doge/modular_doge.py:147
-- shape 2: 1 copy  <- the only one of its shape
   src/transformers/integrations/flex_attention.py:262
```

Doge **forks** the shared flex helper, and the shared one is the minority shape. That single output is what sent me down the flex path, which is where finding #2 below came from. `<- the only one of its shape` is the annotation doing the work: it converts a grouping into a claim.

**`defs` line ranges as load-bearing evidence.** One row decided the session's main review finding:

```
451-452      class     DogeModel
```

Two lines. `modular_doge.py` declares `class DogeModel(MixtralModel):` and stops — so the `create_causal_mask` call the in-flight PR wants to change is *generated* from Mixtral, and cannot be edited in the file modular owns. I did not read either file to establish that. A start-end pair and a class name were enough.

---

## What the session produced that the thread did not

Recorded here because it is the measure of whether the tool paid for itself, not because this issue is the place to fix them.

1. **The fix cannot be written where the PR writes it.** #48769 changes `modeling_doge.py`, which is generated; `modular_doge.py:451` is a two-line stub, so `check_modular_conversion` will reject it.
2. **flex attention is broken too, unconditionally, and the proposed fix does not touch it.** `flex_attention_mask` ignores `allow_is_causal_skip` and always returns a `BlockMask`, which `prepare_dynamic_mask` explicitly skips — so the causal term is dropped for every flex call, padded or not. `_supports_flex_attn = True`.
3. **The slow golden will go red.** `test_Doge_20M_hard` pins `EXPECTED_TEXT` generated under the bug; the PR ran only the non-slow suite.

All three came from `grep`/`symbol`/`copies`/`defs`. None is in the issue or the PR.

---

### 1. `symbol` silently discards an unresolvable qualifier and answers a question I did not ask

This cost me a real detour, and it is the sharpest #11 violation I have hit.

```console
$ relore symbol Axk2Model.forward          # correct spelling is AXK2Model
examples/modular-transformers/modeling_add_function.py:65  method  TestAttention.forward
(8131 definitions of this name; `relore copies` groups them)
    def forward(self) -> tuple[...]:
        _ = apply_rotary_pos_emb(1, 1, 1, 1)
```

Well-formed, confident, complete, and about a different symbol in a different file in a different tree. The class qualifier is dropped when it does not resolve, the final segment is matched alone, and an arbitrary definition is returned with no notice. It is not a near-miss heuristic either:

```console
$ relore symbol ZzzNotARealClass.forward
examples/modular-transformers/modeling_add_function.py:65  method  TestAttention.forward   # same answer
```

The un-dotted path is already correct and is the model for the fix:

```console
$ relore symbol ZzzNotARealSymbol
relore: … 404: ZzzNotARealSymbol is not defined in huggingface/transformers
```

Two aggravating details. The arbitrary pick lands in `examples/modular-transformers/` — a fixture tree no caller wants. And `(8131 definitions of this name; …)` reads as *"your symbol is common"*, when it actually means *"I ignored the half of your query that was selective."*

Suggested behaviour, in the shape of the existing 404: `Axk2Model.forward is not defined in huggingface/transformers; did you mean AXK2Model.forward?` — a `.`-qualified query should never be served by last-segment fallback.

### 2. The `files` source is five days stale, and the disclosure cannot say so

The single most decisive fact about the in-flight PR was *which files it touches* — it is what confirms or kills finding #1 above. relore:

```
huggingface/transformers#48769 pr  open  19h
changed files: none collected of 2, so this thread cannot answer whether it touched a path
```

**The convention worked exactly as #11 requires.** I did not conclude; I wrote "appears to miss" and said so. Good. But the coverage behind it did not, and `status` explains why:

```
huggingface/transformers [files]    high-water None  last-ok 2026-09-09T16:32
huggingface/transformers [threads]  high-water 2026-09-14T10:29  last-ok 2026-09-14T10:47
```

Five days. Every PR opened since has an empty file list by construction. Checked afterwards out-of-band, the two paths were `src/transformers/models/doge/modeling_doge.py` and `tests/models/doge/test_modeling_doge.py` — **no `modular_doge.py`**, which confirms the finding relore had left me unable to confirm.

The filable part is narrow and is not "run the collector more often". It is that `none collected of 2` is ambiguous between three different situations with three different remedies: *this thread has no files*, *the collector has not reached this thread yet*, and *the collector for this repo last succeeded five days ago*. The last one is knowable at render time — `status` already has the row. Something like:

```
changed files: none collected of 2 — the [files] source for huggingface/transformers last succeeded 2026-09-09 (5d),
so this is absence of data, not absence of files
```

That is #11's own argument applied one level up: a disclosure that cannot distinguish *not collected* from *not collected recently* still lets a reader draw the wrong inference about how much to trust it.

### 3. There is no way to read a line range

I worked around this three times. `symbol` gives a definition, `defs` gives ranges, `grep` gives matching lines — but nothing serves *"show me `masking_utils.py:1500-1520"*. That block is a module-level `dict`, so no definition contains it and `defs` will never list it. Getting at it took three calls: guess a regex → `grep` → recognise the name → `defs` to bracket it → `symbol` on the function after it to infer the boundary.

A `relore show PATH:START-END`, or `-C/-A/-B` on `grep`, closes this. `grep` already has the file open. Of the two, context lines on `grep` are the smaller change and cover most of the need, since I always arrived via a match.

### 4. A new datum for #46, pointing the other way

#46 reports `status` *overstating* comment freshness. This session it **understated** it, which is worse for an agent because it invites discounting good data:

```
huggingface/transformers [issue_comments]  high-water 2026-09-09T14:13   # five days old
```

yet `thread 48748` served, correctly, a comment posted 55 minutes earlier:

```
-- 1 of 1 comments, current to 2026-09-14T10:17:02Z --
1. … <1h  issue_comment  @shahd-elgouhary
   > Can I take this issue, please?
```

That comment mattered — it is one of three parties converging on the same bug. I nearly discounted it against the `status` row. So the finding is not "the rows are optimistic", it is **the rows and the `current to` stamps are uncorrelated in both directions**, and the per-response stamp from #32 is the one to trust. If the rows cannot be made to describe what is served, they are more useful labelled as collector telemetry than as freshness.

---

## What carried the session, ranked

1. **`grep` / `symbol` / `copies` / `defs`.** Not as the check on the prose this time — as the primary instrument. Found the precedent, the fork, the stub, and both findings the humans on the thread do not have.
2. **`inflight`.** Third report running, still the highest value-per-call verb in the tool. One call, and the "write a patch" branch died before I opened it.
3. **Scoping annotations.** Four of them changed what I did: `changed files: none collected of 2` stopped a false conclusion, `mentioned in the discussion: 1 — named by somebody, NOT the diff` stopped me treating a filename in prose as a diff, `<- the only one of its shape` produced a finding, and `(from the commit subject, not a staged commit row)` on `why` told me exactly how strong the attribution was.
4. **Machine-tier suppression, verified accurate.** `review: nobody has approved or blocked it` and `0 of 2 comments, 2 machine-tier suppressed` on #48769. Checked out-of-band: zero reviews, and both comments are `github-actions[bot]`. *"No human has looked at this"* is the fact that decides whether to comment on a PR, and relore stated it correctly and compactly.
5. **`thread --full`.** The whole repro script, ungarbled, including the `for impl in ("eager", "sdpa")` loop that is the entire reproduction. #5 stays fixed.
6. **The multi-repo ambiguity error.** `pass repo=: more than one repository is in scope [...] — or set RELORE_REPO to default it for the whole session` — names the problem, the per-call fix, and the session-level fix in one line. Same shape as #19's message, which v3 called the best in the tool.
7. **`why`.** Right answer first try (`#35891`, "Add Doge model" — the bug is original, not a regression). #57's line-exactness caveat did not bite here, but only because I picked the line out of `grep` output, so it was already the semantic one.
8. **`search`.** Honestly the weakest verb this session — see the headline. It was precise when the terms were in the prose (S3 returned exactly the issue and the PR and nothing else) and adjacent-but-unhelpful otherwise. That is the corpus, not the ranker.

`#55` (global flag position) did not bite, because I had it from v7 and never trailed a global. Worth noting that is knowledge, not a fix.

## Filed from this session

Happy to split these out as their own issues in the #9/#10/#11 pattern so they do not close with this write-up:

- `symbol` silently drops an unresolvable class qualifier and returns an arbitrary same-named definition from `examples/` (the #11 pattern; the un-dotted 404 is the model for the fix)
- `changed files: none collected of N` cannot distinguish *no files* from *collector 5 days behind*, though `status` knows which
- no way to read a line range: add `show PATH:START-END`, or `-C/-A/-B` on `grep`
- a `status`/`current to` datapoint for #46, pointing opposite to the one already filed

## Environment

`relore 0.3.6`, client and daemon agreed, `RELORE_API` default, Python 3.10, macOS arm64. Index: 55194 threads / 517518 documents / 435057 raw objects; `huggingface/transformers [threads]` high-water 2026-09-14T10:29Z, `[files]` last-ok 2026-09-09T16:32Z. Quota: 50 → 99 of 5000 for the day, i.e. ~49 server calls for the whole investigation. No checkout, no clone. Agent: Claude Opus 5 via Claude Code, cold on `transformers#48748`, carrying v7's knowledge of relore.

**Comment by tarekziade** [comment]
The defects in this report are now split out so they do not close along with the write-up, following the #9/#10/#11 pattern from #8:

- **#59** — `symbol` drops an unresolvable class qualifier and serves an arbitrary same-named definition out of `examples/`. The #11 pattern, and the un-dotted path already 404s correctly, so the fix has a model in-tree.
- **#60** — `changed files: none collected of N` cannot distinguish *no files* from *the `[files]` collector is 5 days behind*, though `status` knows which. The disclosure worked as designed; this is about the one inference it cannot support.
- **#61** — no way to read a line range: `grep` has no `-C`, and a module-level block is reachable by no verb. Three workarounds in one session.
- **#46** — the `status`/`current to` divergence, [commented there](https://github.com/huggingface/relore/issues/46#issuecomment-5662853164) rather than re-filed, since that issue already notes the superseded rows. The new part is the cost in the opposite direction: it *understated* freshness by five days and nearly made me discard a 55-minute-old comment that was correctly served.

The headline claim of this report — that the code lens is a discovery surface and not only a confirmation surface — is deliberately **not** filed as an issue. It is not a defect and I do not think it implies a change to `TYPICAL ORDER`, which is right for triage. It is an argument about where the code verbs sit in an agent's default reach, and it belongs in the record rather than in a tracker.

One thing I want to keep honest about this session, since every previous report in #8 was cold on the tool and this one was not: I carried v7's knowledge of relore into it. `#55` (global flag position) did not bite me, and I should not be read as evidence that it is less of a problem — I simply never trailed a global, because I already knew not to. The same caveat applies to anything in the ranked list that reads as friction-free.

This issue can be treated as read-only background for the four above.
